analytics/src/analytics/integrations/etldb/issue_model.py

```diff
@@ -0,0 +1,200 @@
+
+        # insert dimensions
+        issue_id = self._insert_dimensions(issue_df, ghid_map)
+        if issue_id is not None:
+            change_type = EtlChangeType.INSERT
+
+        # if insert failed, select and update
+        if issue_id is None:
+            issue_id, change_type = self._update_dimensions(issue_df, ghid_map)
+
+        # insert facts
+        if issue_id is not None:
+            self._insert_facts(issue_id, issue_df, ghid_map)
+
+        return issue_id, change_type
+
+    def _insert_dimensions(self, issue_df: Series, ghid_map: dict) -> int | None:
+        """Write issue dimension data to etl database."""
+        # insert into dimension table: issue
+        new_row_id = None
+        cursor = self.dbh.connection()
+        result = cursor.execute(
+            text(
+                "insert into gh_issue "
+                "(ghid, title, type, opened_date, closed_date, parent_issue_ghid, epic_id) "
+                "values (:ghid, :title, :type, :opened_date, :closed_date, :parent_ghid, :epic_id) "
+                "on conflict(ghid) do nothing returning id",
+            ),
+            {
+                "ghid": issue_df["issue_ghid"],
+                "title": issue_df["issue_title"],
+                "type": issue_df["issue_type"] or "None",
+                "opened_date": issue_df["issue_opened_at"],
+                "closed_date": issue_df["issue_closed_at"],
+                "parent_ghid": issue_df["issue_parent"],
+                "epic_id": ghid_map[EtlEntityType.EPIC].get(issue_df["epic_ghid"]),
+            },
+        )
+        row = result.fetchone()
+        if row:
+            new_row_id = row[0]
+
+        # commit
+        self.dbh.commit(cursor)
+
+        return new_row_id
+
+    def _insert_facts(
+        self,
+        issue_id: int,
+        issue_df: Series,
+        ghid_map: dict,
+    ) -> tuple[int | None, int | None]:
+        """Write issue fact data to etl database."""
+        # get values needed for sql statement
+        issue_df = issue_df.fillna(0)
+        insert_values = {
+            "issue_id": issue_id,
+            "status": issue_df["issue_status"],
+            "is_closed": int(issue_df["issue_is_closed"]),
+            "points": issue_df["issue_points"],
+            "sprint_id": ghid_map[EtlEntityType.SPRINT].get(issue_df["sprint_ghid"]),
+            "effective": self.dbh.effective_date,
+        }
+        history_id = None
+        map_id = None
+
+        # insert into fact table: issue_history
+        cursor = self.dbh.connection()
+        insert_sql1 = text(
+            "insert into gh_issue_history (issue_id, status, is_closed, points, d_effective) "
+            "values (:issue_id, :status, :is_closed, :points, :effective) "
+            "on conflict (issue_id, d_effective) "
+            "do update set (status, is_closed, points, t_modified) = "
+            "(:status, :is_closed, :points, current_timestamp) "
+            "returning id",
+        )
```

Comment:
Oh I see, so this table stores a new record for each effective date and multiple inserts in the same day simply overwrites the previous instance with the same effective date.

Reply:
Yes, the granularity of updates is presumed to be daily